### PR TITLE
feat: download binaries to ansible controller and copy to nodes

### PR DESCRIPTION
**Pull request by gannaramu**
for instances when remote machine dont have internet access.

### CODE REVIEW

**Comment by patrickjahns** [comment]
Thank you very much for your proposal. However - I am very reluctant to accept the change, for the simple reason, that your change now enforces to upload the binaries from the controller for everyone. Imagine a fleet of 100s even 1000s of hosts - the controller would require to copy it to all.

Instead of changing the role, I propose the following:

You create a playbook in which you copy the binaries onto the respective hosts and then specify the paths where they can be found via `{{ promtail_tmp_dir }}` and `{{ promtail_version }}`. Respectively you can also specify the checksum ( see: https://github.com/patrickjahns/ansible-role-promtail/issues/71 // https://github.com/patrickjahns/ansible-role-promtail/pull/72 )

Or alternatively - provide a fileserver as the user in #71 has done

**Comment by gannaramu** [comment]
@patrickjahns I agree with your point. but there are many cases where these nodes are behind firewalls and cant access them from internet. the current logic is inspired from cloudalchemy node exporter ansible role. https://galaxy.ansible.com/cloudalchemy/node_exporter
https://github.com/cloudalchemy/ansible-node-exporter/blob/c2401cfed554906843520cbac7ae2a3a132c590e/tasks/install.yml#L20-L53
 I will leave the decision up to you.

**Comment by patrickjahns** [comment]
Have you tried the suggestion to copy the binary outside of the role? 
If we were to introduce the change, the existing behaviour should not be changed and a new behaviour for fetching binaries and distributing them should be opt-in.

**Comment by gannaramu** [comment]
I didn't get a chance to do it. but i have been using this modified version. I will let you know once i get a change to try it out.

**Comment by patrickjahns** [comment]
@gannaramu 
A similar feature was merged with #104  - maybe this can already be used like you intended it to do?
alternatively that is a good PR to copy and create another feature flag to enable your behaviour - would you pick it up ?

**Comment by gannaramu** [comment]
@patrickjahns sure will try it out and see if it suffices. if not will incorporate it in this pr

**Comment by gannaramu** [comment]
@patrickjahns  what do you think of this implementation?

**Comment by patrickjahns** [comment]
@gannaramu 

Have you tested your idea yet? I fear this won't work as the local path "branch" won't have any checksums available. 
I believe it would be better to hide this download behaviour behind a feature flag to explicitly change the behaviour instead of relying of implicit variables being set.

Please add a molecule test scenario to cover this additional feature

**Comment by gannaramu** [comment]
@patrickjahns can we add support to python3.10 on tox?

**Comment by gannaramu** [comment]
@patrickjahns  i have the molecule tests added need some eyes on it 
https://github.com/gannaramu/ansible-role-promtail/actions/runs/4151862946/jobs/7182434219

File: molecule/download_and_propogate/playbook.yml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+---
+- hosts: all
+  any_errors_fatal: true
+  #become: true
+  roles:
+    - role: ansible-role-promtail
+      vars:
+        promtail_version: "2.7.3"
```
Comment (patrickjahns) [inline molecule/download_and_propogate/playbook.yml:8]: nitpick: I believe we drop the fixed version here

Comment (gannaramu) [inline molecule/download_and_propogate/playbook.yml:8]:
```suggestion
```

File: molecule/download_and_propogate/prepare.yml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  hosts: all
+  gather_facts: true
+  vars:
+    promtail_version: "2.7.3"
+    go_arch_map:
+      x86_64: 'amd64'
+      aarch64: 'arm64'
+      armv7l: 'arm'
+      armv6l: 'arm'
+    go_arch: "{{ go_arch_map[ansible_architecture] | default(ansible_architecture) }}"
```
[the remark this answers is not in the document]
Comment (gannaramu) [reply molecule/download_and_propogate/prepare.yml:12]: I was reffering to an existing file.. https://github.com/patrickjahns/ansible-role-promtail/blob/77c0f6a6f8029b99eecccc4a4dd09af49dd3aac0/molecule/scenario_with_local_binary_directory/playbook.yml#L8

[the remark this answers is not in the document]
Comment (patrickjahns) [reply molecule/download_and_propogate/prepare.yml:12]: That is a special scenario and not the default one. Does your scenario require this to be specified?

[the remark this answers is not in the document]
Comment (patrickjahns) [reply molecule/download_and_propogate/prepare.yml:12]: If there is not use-case that requires special handling, I'd prefer if these values are kept default:
see https://github.com/patrickjahns/ansible-role-promtail/blob/77c0f6a6f8029b99eecccc4a4dd09af49dd3aac0/molecule/default/prepare.yml

[the remark this answers is not in the document]
Comment (gannaramu) [reply molecule/download_and_propogate/prepare.yml]: removed them!

**Review by patrickjahns — APPROVED** [review]

**Comment by patrickjahns** [comment]
@gannaramu - it's finally released with the role 🎉

**Pull request merged**